### ISSUE TITLE
Add LeafNode class

### ISSUE THREAD

**Issue by khaledsulayman** [instructlab/sdg#292]
Currently, leaf node objects are being passed around as a list of dictionaries, but there seems to be a very clear outline [in code](https://github.com/instructlab/sdg/blob/2b73f2b07e266921b30737b10b49c4972d68f1f6/src/instructlab/sdg/utils/taxonomy.py#L174-L197) for what a schema would look like for a LeafNode class.

**Comment by bbrowning** [comment]
Yes, this is a great chance to add standard Python classes or even Pydantic models to stop passing around these dicts everywhere.
